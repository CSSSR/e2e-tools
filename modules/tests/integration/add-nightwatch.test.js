const fs = require('fs')
const path = require('path')
const spawnSync = require('cross-spawn').sync
const { setupEnvironment } = require('./helpers')

function checks({ readFile, rootDir }) {
  it('should add nightwatch with default config to e2e-tools.json', async () => {
    const configFile = JSON.parse(readFile('e2e-tests/e2e-tools.json'))
    expect(configFile.tools).toEqual({
      '@nitive/e2e-tools-nightwatch': {
        browsers: {
          local_chrome: {
            default: true,
            type: 'webdriver',
            desiredCapabilities: {
              browserName: 'chrome',
              'goog:chromeOptions': {
                args: ['--window-size=1200,800'],
              },
            },
            globals: {
              skipScreenshotAssertions: true,
            },
          },
          remote_chrome: {
            type: 'selenium',
            host: 'chromedriver.csssr.ru',
            basicAuth: {
              username_env: 'CHROMEDRIVER_USERNAME',
              password_env: 'CHROMEDRIVER_PASSWORD',
            },

            desiredCapabilities: {
              browserName: 'chrome',
              'goog:chromeOptions': {
                w3c: false,
                args: ['--headless', '--disable-gpu', '--window-size=1200,800'],
              },
            },
            globals: {
              skipScreenshotAssertions: false,
            },
          },
        },
      },
    })
  })

  it('should add package to devDependencies', async () => {
    const packageJson = JSON.parse(readFile('e2e-tests/package.json'))
    expect(packageJson.devDependencies).toHaveProperty('@nitive/e2e-tools-nightwatch')
  })

  it.todo('should add tool with specific version')

  it('should be prettified', () => {
    const { stderr } = spawnSync('yarn', ['prettier', '--check', '**/*.{js,json}'], {
      cwd: path.join(rootDir, 'e2e-tests'),
    })

    const err = stderr && stderr.toString()
    expect(err).toBe('')
  })

  it('should add eslint config file', async () => {
    const eslintConfig = readFile('e2e-tests/nightwatch/.eslintrc.js')
    expect(eslintConfig).toMatchInlineSnapshot(`
      "module.exports = {
        extends: ['@nitive/e2e-tools-nightwatch/eslint'],
      }
      "
    `)
  })

  it('should add .gitignore file', async () => {
    const gitignore = readFile('e2e-tests/nightwatch/.gitignore')
    expect(gitignore).toMatchInlineSnapshot(`
      "report/
      "
    `)
  })

  it('should add screenshots/.gitignore file', async () => {
    const gitignore = readFile('e2e-tests/nightwatch/screenshots/.gitignore')
    expect(gitignore).toMatchInlineSnapshot(`
      "actual/
      diff/
      "
    `)
  })

  it('should add example file', async () => {
    const exist = fs.existsSync(
      path.join(
        rootDir,
        'e2e-tests/nightwatch/tests/Примеры/Переход на страницу авторизации.test.js'
      )
    )
    expect(exist).toBe(true)
  })

  it('should add Jenkinsfile', async () => {
    const jenkinsfile = readFile('e2e-tests/nightwatch/Jenkinsfile')
    expect(jenkinsfile).toMatchSnapshot()
  })

  it('should add Dockerfile', async () => {
    const dockerfile = readFile('e2e-tests/nightwatch/Dockerfile')
    expect(dockerfile).toMatchInlineSnapshot(`
    "FROM node:8

    WORKDIR /usr/src/app

    COPY ../package.json ../yarn.lock ./
    RUN yarn install --frozen-lockfile && yarn cache clean

    COPY . ./nightwatch

    RUN touch .env

    ENTRYPOINT yarn et nightwatch:run --browser remote_chrome
    "
  `)
  })
}

const promptResults = {
  launchUrl: 'github.com',
  projectName: 'github',
  repoSshAddress: 'git@github.com:github/web.git',
}

describe('add-tool command', () => {
  describe('Inside root dir', () => {
    const { run, readFile, rootDir } = setupEnvironment('add-tool-cwd-root')
    run('init')
    run('add-tool @nitive/e2e-tools-nightwatch', { promptResults })
    checks({ readFile, rootDir })
  })

  describe(`Inside e2e-tests dir`, () => {
    const { run, readFile, rootDir } = setupEnvironment('add-tool-cwd-e2e-tests')

    run('init')
    process.chdir(path.join(rootDir, 'e2e-tests'))

    run('add-tool @nitive/e2e-tools-nightwatch', { promptResults })
    checks({ readFile, rootDir })
  })
})
