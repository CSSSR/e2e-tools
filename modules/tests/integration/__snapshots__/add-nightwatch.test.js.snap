// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add-tool command Inside e2e-tests dir should add .gitignore file 1`] = `
"report/
"
`;

exports[`add-tool command Inside e2e-tests dir should add Dockerfile 1`] = `
"FROM node:10

WORKDIR /usr/src/app

COPY package.json yarn.lock ./
RUN yarn install --frozen-lockfile && yarn cache clean

COPY e2e-tools.json ./
COPY nightwatch nightwatch/

RUN touch .env

ENTRYPOINT yarn et nightwatch:run --browser remote_chrome
"
`;

exports[`add-tool command Inside e2e-tests dir should add Jenkinsfile 1`] = `
"pipeline {
  agent any

  environment {
    scmVars = \\"\\"
  }

  parameters {
    string(defaultValue: 'http://github.com', description: 'Адрес стенда', name: 'launch_url', trim: true)
  }

  stages {
    stage('Clone') {
      steps {
        echo \\"Branch: \${GIT_BRANCH}\\"
        script {
          scmVars = checkout([
            $class: 'GitSCM',
            branches: [[name: GIT_BRANCH]],
            doGenerateSubmoduleConfigurations: false,
            extensions: [[$class: 'GitLFSPull']],
            userRemoteConfigs: [[url: 'git@github.com:github/web.git']]
          ])
        }
        echo \\"GIT_BRANCH: \${scmVars.GIT_BRANCH}\\"
        echo \\"GIT_COMMIT: \${scmVars.GIT_COMMIT}\\"
      }
    }

    stage('Install dependencies') {
      steps {
        script {
          sh \\"\\"\\"
            cd e2e-tests
            docker build -f nightwatch/Dockerfile --network host -t github/tests:\${scmVars.GIT_COMMIT} .
          \\"\\"\\"
        }
      }
    }

    stage('Run tests') {
      steps {
        script {
          withCredentials([
            usernamePassword(credentialsId: 'chromedriver', usernameVariable: 'CHROMEDRIVER_USERNAME', passwordVariable: 'CHROMEDRIVER_PASSWORD')
          ]) {
            sh \\"\\"\\"
              docker run --network host \\\\
                -e LAUNCH_URL=\${params.launch_url} \\\\
                -e CHROMEDRIVER_USERNAME \\\\
                -e CHROMEDRIVER_PASSWORD \\\\
                github/tests:\${scmVars.GIT_COMMIT}
            \\"\\"\\"
          }
        }
      }
    }
  }
}
"
`;

exports[`add-tool command Inside e2e-tests dir should add eslint config file 1`] = `
"module.exports = {
  root: true,
  extends: [require.resolve('@nitive/e2e-tools/eslint')],
}
"
`;

exports[`add-tool command Inside e2e-tests dir should add screenshots/.gitignore file 1`] = `
"actual/
diff/
"
`;

exports[`add-tool command Inside root dir should add .gitignore file 1`] = `
"report/
"
`;

exports[`add-tool command Inside root dir should add Dockerfile 1`] = `
"FROM node:10

WORKDIR /usr/src/app

COPY package.json yarn.lock ./
RUN yarn install --frozen-lockfile && yarn cache clean

COPY e2e-tools.json ./
COPY nightwatch nightwatch/

RUN touch .env

ENTRYPOINT yarn et nightwatch:run --browser remote_chrome
"
`;

exports[`add-tool command Inside root dir should add Jenkinsfile 1`] = `
"pipeline {
  agent any

  environment {
    scmVars = \\"\\"
  }

  parameters {
    string(defaultValue: 'http://github.com', description: 'Адрес стенда', name: 'launch_url', trim: true)
  }

  stages {
    stage('Clone') {
      steps {
        echo \\"Branch: \${GIT_BRANCH}\\"
        script {
          scmVars = checkout([
            $class: 'GitSCM',
            branches: [[name: GIT_BRANCH]],
            doGenerateSubmoduleConfigurations: false,
            extensions: [[$class: 'GitLFSPull']],
            userRemoteConfigs: [[url: 'git@github.com:github/web.git']]
          ])
        }
        echo \\"GIT_BRANCH: \${scmVars.GIT_BRANCH}\\"
        echo \\"GIT_COMMIT: \${scmVars.GIT_COMMIT}\\"
      }
    }

    stage('Install dependencies') {
      steps {
        script {
          sh \\"\\"\\"
            cd e2e-tests
            docker build -f nightwatch/Dockerfile --network host -t github/tests:\${scmVars.GIT_COMMIT} .
          \\"\\"\\"
        }
      }
    }

    stage('Run tests') {
      steps {
        script {
          withCredentials([
            usernamePassword(credentialsId: 'chromedriver', usernameVariable: 'CHROMEDRIVER_USERNAME', passwordVariable: 'CHROMEDRIVER_PASSWORD')
          ]) {
            sh \\"\\"\\"
              docker run --network host \\\\
                -e LAUNCH_URL=\${params.launch_url} \\\\
                -e CHROMEDRIVER_USERNAME \\\\
                -e CHROMEDRIVER_PASSWORD \\\\
                github/tests:\${scmVars.GIT_COMMIT}
            \\"\\"\\"
          }
        }
      }
    }
  }
}
"
`;

exports[`add-tool command Inside root dir should add eslint config file 1`] = `
"module.exports = {
  root: true,
  extends: [require.resolve('@nitive/e2e-tools/eslint')],
}
"
`;

exports[`add-tool command Inside root dir should add screenshots/.gitignore file 1`] = `
"actual/
diff/
"
`;
